### ISSUE TITLE
Marten command handler support without requiring Version property

### ISSUE THREAD

**Issue by lofbacke** [JasperFx/wolverine#191]
Hello JasperFx team,

First of all, I want to say how thrilled I am to find the Marten and Wolverine libraries after literally spending years trying to implement all of this functionality myself on a home project. Not only am I glad to see that there are other people trying to do this kind of stuff, but you're doing it _way_ better.

Anyways, after playing with the packages in a sandbox solution for a few hours, I came across an unsupported use case that I'd love to see implemented. It would appear that Wolverine requires a `Version` property to be present in the command class before it will process it. In my existing solution, the client has no concept of stream versioning and the server always assumes that the client wishes to perform the operation upon the latest version of the aggregate. Domain invariants will catch any invalid commands and the event store's "expected version" parameter will catch any concurrency issues during the append operation.

Long story short, it would be great if Wolverine would assume that the command is intended for the current stream version when no explicit `Version` is provided (whether the property be missing entirely, the property is nullable, or the property value is `0`) and fetch the aggregate for writing accordingly.

Thanks again for helping to fill this gaping hole in the community! Your work is appreciated!

**Comment by jeremydmiller** [comment]
@lofbacke I'll double check that because that was all written in a flurry of activity last May, but the Wolverine/Marten combo should fall back to still doing the optimistic concurrency check, but pulling the starting version number from the current state as it starts the action. So you shouldn't have to have the `Version` member. If that's so, I'll adjust the functionality to match what I was describing.

Glad you're liking the concept so far though! @oskardudycz & I have some high hopes for the "critter stack" combo

**Comment by lofbacke** [comment]
@jeremydmiller Thanks for the response! In case you need more information or context, here is the exact code from my sandbox app that I've been testing with. Please forgive my silly aggregate and property names. Maybe I should have stuck with "Foo" and "Bar".

```cs
var builder = WebApplication.CreateBuilder(args);

builder.Services.AddMarten(options =>
{
    options.Connection(builder.Configuration.GetConnectionString("Postgres")
        ?? throw new Exception("Database connection string not found!"));

    if (builder.Environment.IsDevelopment())
    {
        options.AutoCreateSchemaObjects = AutoCreate.All;
    }
})
.IntegrateWithWolverine();

builder.Host.UseWolverine(options =>
{
    options.Handlers.AutoApplyTransactions();
});

var app = builder.Build();

app.MapPost("/thing", async (IDocumentStore store) =>
{
    await using var session = store.OpenSession();

    var thingId = Guid.NewGuid();
    var @event = new ThingStarted(thingId, "Stuff!");
    session.Events.StartStream<Thing>(thingId, @event);
    await session.SaveChangesAsync();

    return thingId;
});

app.MapPut("/thing", (UpdateThing cmd, IMessageBus bus) => bus.InvokeAsync(cmd));

app.Run();

public record ThingStarted(Guid ThingId, string ThingStuff);

public record UpdateThing(Guid ThingId, string StuffToUpdate);

public record ThingUpdated(Guid ThingId, string UpdatedStuff);

public class Thing
{
    public Guid Id { get; set; }
    public int Version { get; set; }
    public string Stuff { get; set; } = string.Empty;

    public void Apply(ThingStarted @event)
    {
        Id = @event.ThingId;
        Stuff = @event.ThingStuff;
    }

    public void Apply(ThingUpdated @event)
    {
        Stuff = @event.UpdatedStuff;
    }
}

public static class UpdateThingAggregateHandler
{
    public static IEnumerable<object> Handle(UpdateThing cmd, Thing aggregate)
    {
        Console.WriteLine($"Loaded {nameof(aggregate)} version {aggregate.Version}");

        yield return new ThingUpdated(cmd.ThingId, cmd.StuffToUpdate);
    }
}
```

And here is the exception that is thrown when I send a request to the PUT endpoint:
```
Exception thrown: 'Wolverine.Runtime.Routing.IndeterminateRoutesException' in Wolverine.dll
An exception of type 'Wolverine.Runtime.Routing.IndeterminateRoutesException' occurred in Wolverine.dll but was not handled in user code: 'Could not determine any valid subscribers or local handlers for message type UpdateThing
System.InvalidOperationException: Compilation failures!

CS1061: 'UpdateThing' does not contain a definition for 'Version' and no accessible extension method 'Version' accepting a first argument of type 'UpdateThing' could be found (are you missing a using directive or an assembly reference?)

Code:

// <auto-generated/>
#pragma warning disable
using Wolverine.Marten.Publishing;

namespace Internal.Generated.WolverineHandlers
{
    // START: UpdateThingHandler987271821
    public class UpdateThingHandler987271821 : Wolverine.Runtime.Handlers.MessageHandler
    {
        private readonly Wolverine.Marten.Publishing.OutboxedSessionFactory _outboxedSessionFactory;

        public UpdateThingHandler987271821(Wolverine.Marten.Publishing.OutboxedSessionFactory outboxedSessionFactory)
        {
            _outboxedSessionFactory = outboxedSessionFactory;
        }



        public override async System.Threading.Tasks.Task HandleAsync(Wolverine.Runtime.MessageContext context, System.Threading.CancellationToken cancellation)
        {
            var updateThing = (UpdateThing)context.Envelope.Message;
            await using var documentSession = _outboxedSessionFactory.OpenSession(context);
            var eventStore = documentSession.Events;
            // Loading Marten aggregate
            var eventStream = await eventStore.FetchForWriting<Thing>(updateThing.ThingId, updateThing.Version, cancellation).ConfigureAwait(false);

            var outgoing1 = UpdateThingAggregateHandler.Handle(updateThing, eventStream.Aggregate);
            if (outgoing1 != null)
            {
                // Capturing any possible events returned from the command handlers
                eventStream.AppendMany(outgoing1);

            }

            await documentSession.SaveChangesAsync(cancellation).ConfigureAwait(false);
        }

    }

    // END: UpdateThingHandler987271821
    
    
}


   at JasperFx.RuntimeCompiler.AssemblyGenerator.Generate(String code)
   at JasperFx.RuntimeCompiler.AssemblyGenerator.Compile(GeneratedAssembly generatedAssembly, IServiceVariableSource services)
   at JasperFx.RuntimeCompiler.CodeFileExtensions.InitializeSynchronously(ICodeFile file, GenerationRules rules, ICodeFileCollection parent, IServiceProvider services)
   at Wolverine.Runtime.Handlers.HandlerGraph.HandlerFor(Type messageType)
   at Wolverine.Runtime.WolverineRuntime.Wolverine.Runtime.IExecutorFactory.BuildFor(Type messageType)
   at Wolverine.Runtime.WolverineRuntime.findInvoker(Type messageType)'
Stack trace:
 >   at Wolverine.Runtime.Handlers.NoHandlerExecutor.InvokeAsync(Object message, MessageBus bus, CancellationToken cancellation, Nullable`1 timeout)
 >   at Wolverine.Runtime.MessageBus.InvokeAsync(Object message, CancellationToken cancellation, Nullable`1 timeout)
 >   at Program.<>c.<<Main>$>b__0_3(UpdateThing cmd, IMessageBus bus) in C:\src\DotNetWebApiTest\DotNetWebApiTest\Program.cs:line 53
 >   at Microsoft.AspNetCore.Http.RequestDelegateFactory.<>c__DisplayClass89_2.<<HandleRequestBodyAndCompileRequestDelegateForJson>b__2>d.MoveNext()
```

Hopefully this helps!

And, yes, I'm loving the concept. As I alluded to in my original post, I've spent the last two years developing my own NuGet packages to do basically the same things as what Marten and Wolverine do. Again, your packages are far more fleshed out and better engineered from what I've gathered. If/when I can finally get my home project to a good place, I would be seriously interested in contributing to these packages. One thing that's been remarkably difficult to find examples of, but that I have spent serious time on in my NuGet packages, are migrations and schema changes in event-sourced systems. I'd love to chat with you and @oskardudycz some more about this particular topic and pick your brains regarding what Marten is already providing and what it might provide in the future.

**Comment by jeremydmiller** [comment]
@lofbacke Sorry for the delay, but I'm looking at this now. And I *never* complain about the problem domain of code for bug reproductions! That's sometimes more work than actually fixing the bug!

**Comment by lofbacke** [comment]
@jeremydmiller Thanks so much for the quick fix!
